File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.js
function solution(phone_book) {
//     const hashTable = new Map();
    
//       for (const number of phone_book) {
//         hashTable[number] = true;
//       };

//       for (const number of phone_book) {
//         for (let i = 1; i < number.length; i++) {
//           const prefix = number.substring(0, i);
//           if (hashTable[prefix]) return false;  
//         };
//       };

//     // 전부 통과했다면 true를 반환한다.
//       return true;

//===================================================
    phone_book.sort();

    const isPrefix = phone_book.some((book,idx)=>{
        return phone_book[idx+1]?.startsWith(book);
    })

    return !isPrefix
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
//         Arrays.sort(phone_book);
        
//         for (int i=0; i<phone_book.length-1; i++) {
//             if (phone_book[i+1].startsWith(phone_book[i])) {
//                 return false;
//             }
//         }
//         return true;
        
        /**
        * HashSet 사용
        **/
        Set<String> set = new HashSet<>(Arrays.asList(phone_book));
        
        for (String number : phone_book) {
            for (int i = 1; i < number.length(); i++) {
                String prefix = number.substring(0, i);
                
                if (set.contains(prefix)) {
                    return false;
                }
            }
        }

        return true;
        
    }
}